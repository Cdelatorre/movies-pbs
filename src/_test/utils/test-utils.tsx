import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
