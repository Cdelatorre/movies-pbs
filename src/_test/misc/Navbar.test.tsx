import { createMemoryHistory } from "@remix-run/router";
import { fireEvent, screen } from "@testing-library/react";
import NavBar from "../../components/misc/navbar";
import { renderWithRouter } from "../utils/test-utils";

describe("Renders Navbar component correctly", () => {
  test("Renders Navbar default component correctly", () => {
    renderWithRouter(<NavBar />);

    const navbar = screen.getByTestId("navbar-component");
    const links: HTMLAnchorElement[] = screen.getAllByRole("link");

    expect(navbar).toBeInTheDocument();
    expect(links[0].textContent).toEqual("");
    expect(links[0].href).toContain("/");

    expect(links[1].textContent).toEqual("Horror");
    expect(links[1].href).toContain(
      `${process.env.REACT_APP_DEV_URL}/?genre=horror`
    );

    expect(links[2].textContent).toEqual("Romance");
    expect(links[2].href).toContain(
      `${process.env.REACT_APP_DEV_URL}/?genre=romance`
    );

    expect(links[3].textContent).toEqual("Comedy");
    expect(links[3].href).toContain(
      `${process.env.REACT_APP_DEV_URL}/?genre=comedy`
    );

    expect(navbar).toMatchSnapshot();
  });

  test("Adds history url on click", () => {
    const history = createMemoryHistory();

    renderWithRouter(<NavBar />, history);

    const navLink = screen.getByRole("link", {
      name: "Romance",
    });

    fireEvent.click(navLink);
    expect(history.location.pathname).toEqual("/");
    expect(history.location.search).toEqual("?genre=romance");
  });
});
