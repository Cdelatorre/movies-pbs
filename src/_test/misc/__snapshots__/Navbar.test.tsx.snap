// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Navbar component correctly Renders Navbar default component correctly 1`] = `
<nav
  class="navbar bg-dark text-light"
  data-testid="navbar-component"
  style="height: 60px;"
>
  <div
    class="container h-100 d-flex align-items-center"
  >
    <a
      aria-current="page"
      class="text-light me-5 active"
      href="/"
    >
      <i
        class="fas fa-home"
        style="font-size: 24px;"
      />
    </a>
    <div
      class="w-100 d-flex align-items-center justify-content-end"
    >
      <a
        aria-current="page"
        class="text-light me-3 ms-5 active"
        href="/?genre=horror"
      >
        Horror
      </a>
      <a
        aria-current="page"
        class="text-light me-3 active"
        href="/?genre=romance"
      >
        Romance
      </a>
      <a
        aria-current="page"
        class="text-light me-3 active"
        href="/?genre=comedy"
      >
        Comedy
      </a>
    </div>
  </div>
</nav>
`;
